core/src/main/scala/org/apache/spark/CacheManager.scala

```diff
@@ -47,9 +49,13 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
         val inputMetrics = blockResult.inputMetrics
         val existingMetrics = context.taskMetrics
           .getInputMetricsForReadMethod(inputMetrics.readMethod)
-        existingMetrics.addBytesRead(inputMetrics.bytesRead)
+        existingMetrics.incBytesRead(inputMetrics.bytesRead)
 
-        new InterruptibleIterator(context, blockResult.data.asInstanceOf[Iterator[T]])
+        val iter = blockResult.data.asInstanceOf[Iterator[T]]
+        new InterruptibleIterator(context, AfterNextInterceptingIterator(iter, (next: T) => {
+          existingMetrics.incRecordsRead(1)
```

Comment:
@sryza right. So how do you propose we increment the bytes and records read in a threadsafe way? If we use a @volatile Long we can't safely do an increment unless we guarantee that only one thread is accessing the InputMetrics at any one time. I guess this is an okay assumption now but doesn't that open ourselves up to race conditions down the line when we add more multithreading? 

Looking at: http://stackoverflow.com/questions/2538070/atomic-operation-cost it doesn't seem like the cost of CAS is that high, there is at most 2 cacheline misses for this integer and only 1 if other CPUs are not reading and writing from it.  Am i misinterpreting this?